node/worker/worker.go

```diff
@@ -92,107 +96,97 @@ func (w *Worker) throttleTxs(selected types.Transactions, recentTxsStats types.R
-	unselected := types.Transactions{}
-	invalid := types.Transactions{}
-	for _, tx := range txs {
-		if tx.ShardID() != w.chain.ShardID() {
-			invalid = append(invalid, tx)
-			continue
-		}
+	txs := types.NewTransactionsByPriceAndNonce(w.current.signer, pendingNormal)
+
+	var coalescedLogs []*types.Log
 
+	// NORMAL
+	for {
 		// If we don't have enough gas for any further transactions then we're done
 		if w.current.gasPool.Gas() < params.TxGas {
-			utils.Logger().Info().Str("Not enough gas for further transactions, have", w.current.gasPool.String()).Uint64("want", params.TxGas)
-			unselected = append(unselected, tx)
+			utils.Logger().Info().Uint64("have", w.current.gasPool.Gas()).Uint64("want", params.TxGas).Msg("Not enough gas for further transactions")
+			break
+		}
+		// Retrieve the next transaction and abort if all done
+		tx := txs.Peek()
+		if tx == nil {
+			break
+		}
+		// Error may be ignored here. The error has already been checked
+		// during transaction acceptance is the transaction pool.
+		//
+		// We use the eip155 signer regardless of the current hf.
+		from, _ := types.Sender(w.current.signer, tx)
+		// Check whether the tx is replay protected. If we're not in the EIP155 hf
+		// phase, start ignoring the sender until we do.
+		if tx.Protected() && !w.config.IsEIP155(w.current.header.Number()) {
+			utils.Logger().Info().Str("hash", tx.Hash().Hex()).Str("eip155Epoch", w.config.EIP155Epoch.String()).Msg("Ignoring reply protected transaction")
+
+			txs.Pop()
 			continue
 		}
+		// Start executing the transaction
+		w.current.state.Prepare(tx.Hash(), common.Hash{}, len(w.current.txs))
 
-		sender, flag := w.throttleTxs(selected, recentTxsStats, txsThrottleConfig, tx)
-		switch flag {
-		case shardingconfig.TxUnselect:
-			unselected = append(unselected, tx)
-
-		case shardingconfig.TxInvalid:
-			invalid = append(invalid, tx)
-
-		case shardingconfig.TxSelect:
-			if tx.GasPrice().Uint64() == 0 {
-				invalid = append(invalid, tx)
-			} else {
-				snap := w.current.state.Snapshot()
-				_, err := w.commitTransaction(tx, coinbase)
-				if err != nil {
-					w.current.state.RevertToSnapshot(snap)
-					invalid = append(invalid, tx)
-					utils.Logger().Error().Err(err).Str("txId", tx.Hash().Hex()).Msg("Commit transaction error")
-				} else {
-					selected = append(selected, tx)
-					// handle the case when msg was not able to extracted from tx
-					if len(sender.String()) > 0 {
-						recentTxsStats[newBlockNum][sender]++
-					}
-				}
-			}
+		if tx.ShardID() != w.chain.ShardID() {
+			txs.Shift()
+			continue
 		}
 
-		// log invalid or unselected txs
-		if flag == shardingconfig.TxUnselect || flag == shardingconfig.TxInvalid {
-			utils.Logger().Info().Str("txId", tx.Hash().Hex()).Str("txThrottleFlag", flag.String()).Msg("Transaction Throttle flag")
+		logs, err := w.commitTransaction(tx, coinbase)
+		sender, _ := common2.AddressToBech32(from)
+		switch err {
+		case core.ErrGasLimitReached:
```

Comment:
we are doing a double check for ErrGasLimitReached. we should remove the gas limit check above and have it failed here for modularity.

Reply:
it's not the same and this logic is following what ethereum is doing. The first check is block level check to say that there is no way any more txn can be added. This current check is for a specific txn. If the current txn is big and it may fail the check, but it doesn't mean the next txn won't succeed.